File: Mue.Clients.Common/ClientServer/HubClient.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;
using Mue.Common.ClientServer;
using Mue.Common.Models;

namespace Mue.Clients.ClientServer
{
    public class HubClient : IClientCommands, IDisposable
    {
        private readonly HubConnection _hub;
        private readonly IServerCommands _handler;
        private readonly List<IDisposable> _disposers = new List<IDisposable>();

        public HubClient(HubConnection hub, IServerCommands handler)
        {
            _hub = hub;
            _handler = handler;

            ConfigureServerCommands();
        }

        public void Dispose()
        {
            if (_disposers != null)
            {
                _disposers.ForEach(f => f.Dispose());
            }
        }

        /// <summary>Client intent: I am requesting authentication.</summary>
        public Task<OperationResponse> Auth(AuthRequest data) => _hub.InvokeAsync<OperationResponse>("Auth", data);
        /// <summary>Client intent: I am sending a command for the server to execute.</summary>
        public Task<OperationResponse> Command(CommandRequest request) => _hub.InvokeAsync<OperationResponse>("Command", request);
        /// <summary>Client intent: Echo test.</summary>
        public Task<OperationResponse> Echo(string message) => _hub.InvokeAsync<OperationResponse>("Echo", message);
        /// <summary>Client intent: I am going to close the session.</summary>
        public Task<OperationResponse> Disconnect() => _hub.InvokeAsync<OperationResponse>("Disconnect");

        private void ConfigureServerCommands()
        {
            _disposers.AddRange(new List<IDisposable> {
                _hub.On<string>("Welcome", _handler.Welcome),
                // _hub.On<AuthResponse, MueCodes>("AuthResponse", _handler.AuthResponse),
                _hub.On<CommunicationsMessage, MueCodes>("Message", _handler.Message),
                _hub.On<string>("Echo", _handler.Echo),
                _hub.On<string>("Disconnect", _handler.Disconnect),
                _hub.On<string, MueCodes>("Fatal", _handler.Fatal),
            });
        }

    }
}

File: Mue.Clients.Web/src/app/mue/HubClient.ts
import * as signalR from "@microsoft/signalr";
import {
    AuthRequest,
    CommandRequest,
    CommunicationsMessage,
    MueCodes,
    OperationResponse,
} from "./consts";

export class HubClient {
    private _isAuthenticated: boolean = false;
    private _cachedAuth?: AuthRequest;

    constructor(
        protected connection: signalR.HubConnection,
        callbacks: HubClientCallbacks,
    ) {
        connection.on("Welcome", (motd: string) => {
            if (this._cachedAuth) {
                this.auth(this._cachedAuth);
                callbacks._internal("Reconnected to server");
            } else {
                this._isAuthenticated = false;
                callbacks.onWelcome(motd);
            }
        });
        connection.on("Message", callbacks.onMessage);
        connection.on("Echo", callbacks.onEcho);
        connection.on("Disconnect", callbacks.onDisconnect);
        connection.on("Fatal", callbacks.onFatal);
    }

    get isAuthenticated() {
        return this._isAuthenticated;
    }

    // client to server
    auth = async (data: AuthRequest) => {
        const res = await this.connection.invoke<OperationResponse>(
            "Auth",
            data,
        );

        if (res.code === MueCodes.Success) {
            this._isAuthenticated = true;
            this._cachedAuth = data;
        } else {
            this._isAuthenticated = false;
        }

        return res;
    };
    command = (request: CommandRequest) =>
        this.connection.invoke<OperationResponse>("Command", request);
    echo = (message: string) =>
        this.connection.invoke<OperationResponse>("Echo", message);
    disconnect = () => this.connection.invoke<OperationResponse>("Disconnect");
}

export interface HubClientCallbacks {
    onWelcome: (motd: string) => void;
    onMessage: (message: CommunicationsMessage, code: number) => void;
    onEcho: (message: string) => void;
    onDisconnect: (reason?: string) => void;
    onFatal: (message: string, code: number) => void;
    _internal: (message: string) => void;
}
